library/src/main/java/com/getstream/sdk/chat/adapter/MessageViewHolderFactory.kt

```diff
@@ -0,0 +1,103 @@
+package com.getstream.sdk.chat.adapter
+
+import android.view.ViewGroup
+import com.getstream.sdk.chat.R
+import com.getstream.sdk.chat.adapter.MessageListItem.DateSeparatorItem
+import com.getstream.sdk.chat.adapter.MessageListItem.MessageItem
+import com.getstream.sdk.chat.adapter.MessageListItem.ThreadSeparatorItem
+import com.getstream.sdk.chat.adapter.MessageListItem.TypingItem
+import com.getstream.sdk.chat.model.ModelType
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.models.Message
+
+/**
+ * Allows you to easily customize message rendering or message attachment rendering
+ */
+open class MessageViewHolderFactory {
+    companion object {
+        const val MESSAGEITEM_DATE_SEPARATOR = 1
+        const val MESSAGEITEM_MESSAGE = 2
+        const val MESSAGEITEM_TYPING = 3
+        const val MESSAGEITEM_THREAD_SEPARATOR = 4
+        const val MESSAGEITEM_NOT_FOUND = 5
+
+        const val GENERIC_ATTACHMENT = 1
+        const val IMAGE_ATTACHMENT = 2
+        const val VIDEO_ATTACHMENT = 3
+        const val FILE_ATTACHMENT = 4
+    }
+
+    open fun getMessageViewType(messageListItem: MessageListItem?): Int {
+        return when (messageListItem) {
+            is DateSeparatorItem -> MESSAGEITEM_DATE_SEPARATOR
+            is TypingItem -> MESSAGEITEM_TYPING
+            is MessageItem -> MESSAGEITEM_MESSAGE
+            is ThreadSeparatorItem -> MESSAGEITEM_THREAD_SEPARATOR
+            else -> MESSAGEITEM_NOT_FOUND
+        }
+    }
+
+    open fun getAttachmentViewType(
+        message: Message,
+        mine: Boolean,
+        position: Position,
+        attachments: List<Attachment>,
+        attachment: Attachment
+    ): Int {
+        return when (attachment.type) {
+            null ->
+                GENERIC_ATTACHMENT
+            ModelType.attach_video ->
+                VIDEO_ATTACHMENT
+            ModelType.attach_image, ModelType.attach_giphy ->
+                IMAGE_ATTACHMENT
+            ModelType.attach_file ->
+                FILE_ATTACHMENT
+            else ->
+                GENERIC_ATTACHMENT
+        }
+    }
+
+    open fun createMessageViewHolder(adapter: MessageListItemAdapter, parent: ViewGroup, viewType: Int): BaseMessageListItemViewHolder<*>? {
+        return when (viewType) {
+            MESSAGEITEM_DATE_SEPARATOR ->
+                DateSeparatorViewHolder(R.layout.stream_item_date_separator, parent)
+            MESSAGEITEM_MESSAGE ->
+                MessageListItemViewHolder(R.layout.stream_item_message, parent).apply {
+                    setMessageClickListener(adapter.messageClickListener)
```

Comment:
What do you think about to decouple our `ViewHolder` from our `Adapter`?
As I see, all logic on this case is coupled.
`Factory` doesn't need to know about the `Adapter`. `Adapter` has some "listeners" that are set only to be send to the `factory` to instanciate the different `ViewHolders`.

At the end, the only one that need the listener is the `factory` class to be able to generate our `ViewHolder`.
It can be added to the `factory` constructor (If we want them mandatory) and try to remove them from `Adapter` class

Reply:
We need more than half a dozen listeners here in the factory, to be able to create `MessageListItemViewHolder` instances. Passing all of those in separately seems impractical. If we want to decouple from the adapter, we could hide it behind an interface, something like a `ListenerProvider` (only with a better name).

One thing to note is that custom implementations of this factory, as it is now, could be using the adapter reference for more than just the click listeners it has. This, especially with the changes in https://github.com/GetStream/stream-chat-android/pull/617 includes the current channel, the bubble helper, the style, whether we're in a thread... Do we want to allow access to these things in the factory?

Reply:
I like the idea to have something like a `ListenerProvider` if we have a lot of listeners to be configured. It can be an interface that belongs to our Factory. If some listener is not mandatory, it could have a default implementation that does nothing, it will help dev to use it.

About the use of the adapter into our factory, in the future, it should be removed, in the end, the factory is who provides stuff to the adapter and not in the other direction. If it is hard to remove it now, we can iterate over it in another task.

Lastly, about the BubbleHelper, Style... I don't think it is needed into our factory, but if we would need it, it should be possible to add to the factory as another dependency on the constructor, keeping methods simple as possible

Reply:
My thoughts about this. Here we want to create an instance of ViewHolder, but listeners are required to setup it. We could decouple this two operations and provide functional object that know how to setup particular type of ViewHolder. It could have a such interface:

```kotlin
interface ViewHolderSetupHandler /* whatever name :-) */ {
  fun setup(viewHolder: BaseMessageListItemViewHolder)
}
```

And we can provide it via constructor. It is more flexible than providing listeners